tools/chapel-py/src/resolution.cpp

```diff
@@ -288,3 +290,111 @@ std::vector<int> const& actualOrderForNode(Context* context, const AstNode* node
+    // Skip non-user functions.
+    if (parsing::idIsInBundledModule(context, fn->id())) return false;
+
+    using namespace resolution;
+
+    // The function also needs to be concrete.
+    ResolutionContext rcval(context);
+    const UntypedFnSignature* uSig = UntypedFnSignature::get(context, fn);
+    const TypedFnSignature* sig = typedSignatureInitial(&rcval, uSig);
+    if (sig == nullptr || sig->needsInstantiation()) return false;
+
+    if (canPass(context, testType, sig->formalType(0)).passes()) {
+      // One formal of 'testType', which constitutes a test.
+      fns.push_back(fn);
+    }
+
+    return false;
+  }
+  void exit(const Function* fn) {}
+
+  bool enter(const AstNode* node) { return true; }
+  void exit(const AstNode* node) {}
+};
+
+std::vector<const Function*> const&
+findTestFunctionsForModule(Context* context, const Module* mod) {
+  QUERY_BEGIN(findTestFunctionsForModule, context, mod);
+  std::vector<const Function*> result;
+
+  using namespace parsing;
+
+  // make sure UnitTest has been parsed before getSymbolIdFromTopLevelModule
+  if (getToplevelModule(context, UniqueString::get(context, "UnitTest"))) {
+    if (auto TestID = getSymbolIdFromTopLevelModule(context, "UnitTest", "Test")) {
+      auto TestTy = resolution::initialTypeForTypeDecl(context, TestID);
+      if (TestTy && TestTy->isClassType()) {
+        auto borrowedDec = ClassTypeDecorator(ClassTypeDecorator::BORROWED_NONNIL);
+        auto borrowedTestTy = TestTy->toClassType()->withDecorator(context, borrowedDec);
+        auto borrowedTestQualTy = QualifiedType(QualifiedType::DEFAULT_INTENT, borrowedTestTy);
+
+        TestFunctionFinder tff(context, borrowedTestQualTy);
+        mod->traverse(tff);
+        result = tff.fns;
+      }
+    }
+  }
+  return QUERY_END(result);
+}
+
+struct TestFunctionMainFinder {
+  // Inputs
+  Context* context;
+  ID mainID;
+
+  // Outputs
+  const FnCall* callsMain;
+
+  TestFunctionMainFinder(Context* context, ID mainID)
+    : context(context), mainID(mainID), callsMain(nullptr) {}
+
+  bool enter(const FnCall* call) {
+    if (callsMain != nullptr) return false;
+    auto resolvedID = scopeResolveToIdForNode(context, call->calledExpression());
+    if (resolvedID == mainID) {
+      callsMain = call;
+      return false;
+    }
+    return true;
+  }
+  void exit(const Function* fn) {}
+
+  bool enter(const AstNode* node) { return callsMain == nullptr; }
+  void exit(const AstNode* node) {}
+};
+
+const FnCall* const&
+findUnitTestMainForModule(Context* context, const Module* mod) {
+  QUERY_BEGIN(findUnitTestMainForModule, context, mod);
+  const FnCall* result = nullptr;
+
+  using namespace parsing;
+
+  // make sure UnitTest has been parsed before getSymbolIdFromTopLevelModule
+  if (getToplevelModule(context, UniqueString::get(context, "UnitTest"))) {
+    if (auto mainID = getSymbolIdFromTopLevelModule(context, "UnitTest", "main")) {
```

Comment:
You've re-used this condition twice. Consider making a new helper.

Reply:
It seems like overkill to make a helper for a single line

Reply:
done in https://github.com/chapel-lang/chapel/pull/27706/commits/f3e980d81365bef88f025eb2591189793be2e16e